[SPARK-44818][CORE] Fix race for pending task kill issued before taskThread is initialized

anishshri-db opened the pull request:
### What changes were proposed in this pull request?
Fix race for pending task kill issued before taskThread is initialized

### Why are the changes needed?
We see that there is a race for tasks that are interrupted through stage cancellation and that may be added to the TaskSet, but don't yet have taskThread initialized.

Basically, we try to kill ongoing task attempts to handle stage cancellation

```
    logInfo("Cancelling stage " + stageId)
    // Kill all running tasks for the stage.
    killAllTaskAttempts(stageId, interruptThread, reason = "Stage cancelled: " + reason)
    // Cancel all attempts for the stage.
```

However, there is a chance that taskThread is not initialized yet and we only set the reasonIfKilled.

```
  def kill(interruptThread: Boolean, reason: String): Unit = {
    require(reason != null)
    _reasonIfKilled = reason
    if (context != null) {
      context.markInterrupted(reason)
    }
    if (interruptThread && taskThread != null) {
      taskThread.interrupt().  <--- never hit
    }
```

Then within the task execution thread itself, we try to call kill again since the reasonIfKilled is set. However, this time we pass interruptThread as false explicitly since we don't know the status of the previous call.

```
    taskThread = Thread.currentThread()

    if (_reasonIfKilled != null) {
      kill(interruptThread = false, _reasonIfKilled) <-- only context will be set, 
    }
```

The TaskReaper has also finished its previous and only attempt at task interruption since we don't try for multiple times in this case. Eventually, the task is not interrupted even once and it gets blocked on some I/O or wait calls which might not finish within the reaper timeout, leading to the JVM being killed.

```
        taskRunner.kill(interruptThread = interruptThread, reason = reason)
```

The change tries to fix this issue by checking for the presence of `reasonIfKilled` on the context and issuing a `TaskKilledException` before we execute `runTask` thereby preventing execution of the actual task and freeing up the slot and also preventing future issues with the reaper.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing unit tests

```
[info] JobCancellationSuite:
...
[info] Run completed in 35 seconds, 781 milliseconds.
[info] Total number of tests run: 13
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 13, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

anishshri-db committed:
[SPARK-44818] Fix race for pending interrupt issued before taskThread is initialized

anishshri-db commented:
@jiangxb1987 @HeartSaVioR - PTAL, thx !

mridulm commented:
+CC @JoshRosen

anishshri-db commented:
@jiangxb1987 @JoshRosen - could you folks please take a look ? Thx

JoshRosen reviewed (commented):
I agree that there is a race condition here, but I wonder whether there might potentially be a simpler fix:

- It seems like the net effect of this change is to have the `kill(interruptThread = false, _reasonIfKilled)` call perform an interrupt of `taskThread` if the earlier kill request would have issued an interrupt. When that second `kill` call is invoked, `taskThread == Thread.currentThread()`, so the task thread interrupts itself. Because it's a self-interrupt, this implies that the task isn't blocked in IO when the interrupt is issued. Therefore, the interrupt issued at that point must have the net effect of interrupting some _future_ IO operation.
- At the point where that second `kill` call is issued, `context` has already been assigned, so that `kill` is calling `context.markInterrupted(reason)`. There's many places in Spark where code checks `TaskContext.isInterrupted` or `TaskContext.killTaskIfInterrupted()`, but if tasks are hanging due to the race condition then there must be cases where tasks can get stuck in IO prior to any checks for whether the task is interrupted.
- As an alternative fix, what if we modified `TaskContext.runTaskWithListeners()` to add a `killTaskIfInterrupted()` right before we invoke the task run method? i.e. in https://github.com/apache/spark/blob/48faaa8ee73d8005d2ed0668b4d0e860fc92ca4d/core/src/main/scala/org/apache/spark/TaskContext.scala#L159-L161 add a  `killTaskIfInterrupted()` as the first line of the `try`?  This ensures that all tasks will first check whether they have been canceled before they begin executing.
  -  I think this resolves the race in case where an interrupting task kill is issued before the task thread is launched: the task will start and then immediately self-destruct.
  -  I think my suggestion is slightly simpler / easy to reason about (but that's a bit subjective).
    - This approach reduces (but does not entirely eliminate) the risk that an InterruptedException is thrown from `Task.run()` _before_ the `try` block. That might happen if we had an executor plugin that performed IO in `    plugins.foreach(_.onTaskStart())`, for example. I haven't fully analyzed whether we have another pre-existing problem that needs fixing here (e.g. maybe by shifting more stuff into that `try`?), but avoiding new interrupt sources at least doesn't _worsen_ any potential pre-existing problems.

WDYT?

anishshri-db committed:
Change to address Josh's comments

anishshri-db committed:
Remove added line

anishshri-db commented:
@JoshRosen - I addressed your comments. I think your proposal seems fine too. At the time that we throw this TaskKilled exception, the context should be non-null and should have the reason set even though interruptThread is passed as false. The issue of whether the plugins will block on I/O remains. We should possibly move that within this `try` after the check for killIfInterrupted maybe ? But in any case, the issues we have seen have been with threads that actually run the task execution and get blocked on some network I/O (remote RPC or otherwise) and whose timeouts are effectively larger than the reaper timeout causing us to block task slots and force a executor JVM kill eventually. Those majority of cases should be handled by your proposed fix, I believe. Let me know what you think. Thx

JoshRosen reviewed (approved):
> The issue of whether the plugins will block on I/O remains. We should possibly move that within this try after the check for killIfInterrupted maybe.
>
> But in any case, the issues we have seen have been with threads that actually run the task execution and get blocked on some network I/O (remote RPC or otherwise) and whose timeouts are effectively larger than the reaper timeout causing us to block task slots and force a executor JVM kill eventually. Those majority of cases should be handled by your proposed fix, I believe. Let me know what you think.

Agreed: I think this cleanly handles the likely cases in the non-plugin scenario.

I'm fine with leaving the plugin code as-is until we have evidence that problems are occurring there in practice. I'd bet that the combination of "stage cancellation + task reaper + executor plugin + slow IO in executor plugin" is sufficiently rare that there's not a strong motivation to guard against edge cases there, so I'm fine erring on the side of caution by _not_ making changes there (thereby avoiding risks of unanticipated side effects, etc).

---

Given this, LGTM on the current change. 

Do you mind updating the PR description to reflect the final changes?

anishshri-db commented:
> I'm fine with leaving the plugin code as-is until we have evidence that problems are occurring there in practice. I'd bet that the combination of "stage cancellation + task reaper + executor plugin + slow IO in executor plugin" is sufficiently rare that there's not a strong motivation to guard against edge cases there, so I'm fine erring on the side of caution by not making changes there (thereby avoiding risks of unanticipated side effects, etc).
Yup agreed.

> Do you mind updating the PR description to reflect the final changes?
Done

anishshri-db commented:
@JoshRosen @HeartSaVioR - could one of you pls merge when you get a chance ? Thx

JoshRosen commented:
Merged to master. Thanks @anishshri-db!
